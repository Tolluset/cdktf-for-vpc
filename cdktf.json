{
  "language": "typescript",
  "app": "npx ts-node main.ts",
  "projectId": "772deac9-ed06-47ad-bbe5-25cbbc62ae0b",
  "sendCrashReports": "false",
  "terraformProviders": [],
  "terraformModules": [],
  "context": {
    "excludeStackIdFromLogicalIds": "true",
    "allowSepCharsInLogicalIds": "true"
  }
}