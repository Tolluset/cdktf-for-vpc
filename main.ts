import { Construct, IConstruct } from "constructs";
import { App, TerraformStack, Aspects, IAspect } from "cdktf";
import {
  vpc,
  provider,
  subnet,
  internetGateway,
  routeTable,
  route,
  routeTableAssociation,
} from "@cdktf/provider-aws";

const NOW = new Date().toLocaleString("ko-KR", { timeZone: "Asia/Seoul" });

type TaggableConstruct = IConstruct & {
  tags?: { [key: string]: string };
  tagsInput?: { [key: string]: string };
};

function isTaggableConstruct(x: IConstruct): x is TaggableConstruct {
  return "tags" in x && "tagsInput" in x;
}

class Tagged implements IAspect {
  constructor(private tags: Record<string, string>) {}

  visit(node: IConstruct): void {
    if (isTaggableConstruct(node)) {
      const currentTags = node.tagsInput || {};
      node.tags = {
        Name: node.node.id,
        CreatedAt: NOW,
        ...this.tags,
        ...currentTags,
      };
    }
  }
}

class VpcStack extends TerraformStack {
  constructor(scope: Construct, id: string) {
    super(scope, id);

    new provider.AwsProvider(this, `${id}-aws-provider`, {
      region: "ap-northeast-2",
    });

    const newVpc = new vpc.Vpc(this, `${id}-vpc`, {
      cidrBlock: "10.0.0.0/16",
    });

    const newInternetGateway = new internetGateway.InternetGateway(
      this,
      `${id}-internet-gateway`,
      {
        vpcId: newVpc.id,
      }
    );

    const publicSubnetA = new subnet.Subnet(this, `${id}-public-subnet-a`, {
      vpcId: newVpc.id,
      availabilityZone: "ap-northeast-2a",
      mapPublicIpOnLaunch: true,
      cidrBlock: "10.0.1.0/24",
    });

    const publicSubnetC = new subnet.Subnet(this, `${id}-public-subnet-c`, {
      vpcId: newVpc.id,
      availabilityZone: "ap-northeast-2c",
      mapPublicIpOnLaunch: true,
      cidrBlock: "10.0.3.0/24",
    });

    const publicRouteTable = new routeTable.RouteTable(
      this,
      `${id}-public-route-table`,
      {
        vpcId: newVpc.id,
      }
    );

    new routeTableAssociation.RouteTableAssociation(
      this,
      `${id}-route-table-association-public-subnet-a`,
      {
        routeTableId: publicRouteTable.id,
        subnetId: publicSubnetA.id,
      }
    );

    new routeTableAssociation.RouteTableAssociation(
      this,
      `${id}-route-table-association-public-subnet-c`,
      {
        routeTableId: publicRouteTable.id,
        subnetId: publicSubnetC.id,
      }
    );

    new route.Route(this, `${id}-route-public-to-internet-gateway`, {
      destinationCidrBlock: "0.0.0.0/0",
      routeTableId: publicRouteTable.id,
      gatewayId: newInternetGateway.id,
    });

    const privateSubnetA = new subnet.Subnet(this, `${id}-private-subnet-a`, {
      vpcId: newVpc.id,
      availabilityZone: "ap-northeast-2a",
      cidrBlock: "10.0.101.0/24",
    });

    const privateSubnetC = new subnet.Subnet(this, `${id}-private-subnet-c`, {
      vpcId: newVpc.id,
      availabilityZone: "ap-northeast-2c",
      cidrBlock: "10.0.103.0/24",
    });

    const privateRouteTable = new routeTable.RouteTable(
      this,
      `${id}-private-route-table`,
      {
        vpcId: newVpc.id,
      }
    );

    new routeTableAssociation.RouteTableAssociation(
      this,
      `${id}-route-table-association-private-subnet-a`,
      {
        routeTableId: privateRouteTable.id,
        subnetId: privateSubnetA.id,
      }
    );

    new routeTableAssociation.RouteTableAssociation(
      this,
      `${id}-route-table-association-private-subnet-c`,
      {
        routeTableId: privateRouteTable.id,
        subnetId: privateSubnetC.id,
      }
    );

    Aspects.of(this).add(new Tagged({ CreatedBy: "cdktf", Project: id }));
  }
}

const app = new App();

new VpcStack(app, "cdktf-for-vpc");

app.synth();
