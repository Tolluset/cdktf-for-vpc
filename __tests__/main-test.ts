import "cdktf/lib/testing/adapters/jest";
import { TerraformStack, Testing } from "cdktf";
import {
  vpc,
  provider,
  subnet,
  internetGateway,
  routeTable,
} from "@cdktf/provider-aws";
import { Tagged, VpcStack } from "../main";

const NOW = "2023. 3. 31. 오전 11:43:10";
const slots = [
  new Tagged({ CreatedBy: "cdktf", Project: "cdktf-for-vpc" }, NOW),
  new Tagged({ how: "it", Project: "cdktf-for-vpc" }, NOW),
];

describe("Vpc stack", () => {
  it("should contain the follow resources", () => {
    const app = Testing.app();
    const stack = new VpcStack(app, "test", slots);
    const synthesized = Testing.synth(stack);

    expect(synthesized).toHaveProviderWithProperties(
      provider.AwsProvider,
      {
        region: "ap-northeast-2",
      }
    );
    expect(synthesized).toHaveResource(vpc.Vpc);
    expect(synthesized).toHaveResource(subnet.Subnet);
    expect(synthesized).toHaveResource(
      internetGateway.InternetGateway
    );
    expect(synthesized).toHaveResource(routeTable.RouteTable);
  });

  it("compare snapshots", () => {
    const app = Testing.app();
    const stack = new VpcStack(app, "test", slots);

    expect(Testing.synth(stack)).toMatchSnapshot();
  });

  it("to be valid terraform stack", () => {
    const app = Testing.app();
    const stack = new TerraformStack(app, "test");

    new VpcStack(stack, "test-app", slots);

    expect(Testing.fullSynth(stack)).toBeValidTerraform();
  });

  it("to be planned", () => {
    const app = Testing.app();
    const stack = new TerraformStack(app, "test");

    new VpcStack(stack, "test-app", slots);

    expect(Testing.fullSynth(stack)).toPlanSuccessfully();
  });
});
