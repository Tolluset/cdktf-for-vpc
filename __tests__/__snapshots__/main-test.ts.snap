// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Vpc stack compare snapshots 1`] = `
"{
  "provider": {
    "aws": [
      {
        "region": "ap-northeast-2"
      }
    ]
  },
  "resource": {
    "aws_internet_gateway": {
      "test-internet-gateway": {
        "tags": {
          "CreatedBy": "cdktf",
          "Name": "test-internet-gateway",
          "Project": "cdktf-for-vpc",
          "how": "it"
        },
        "vpc_id": "\${aws_vpc.test-vpc.id}"
      }
    },
    "aws_route": {
      "test-route-public-to-internet-gateway": {
        "destination_cidr_block": "0.0.0.0/0",
        "gateway_id": "\${aws_internet_gateway.test-internet-gateway.id}",
        "route_table_id": "\${aws_route_table.test-public-route-table.id}"
      }
    },
    "aws_route_table": {
      "test-private-route-table": {
        "tags": {
          "CreatedBy": "cdktf",
          "Name": "test-private-route-table",
          "Project": "cdktf-for-vpc",
          "how": "it"
        },
        "vpc_id": "\${aws_vpc.test-vpc.id}"
      },
      "test-public-route-table": {
        "tags": {
          "CreatedBy": "cdktf",
          "Name": "test-public-route-table",
          "Project": "cdktf-for-vpc",
          "how": "it"
        },
        "vpc_id": "\${aws_vpc.test-vpc.id}"
      }
    },
    "aws_route_table_association": {
      "test-route-table-association-private-subnet-a": {
        "route_table_id": "\${aws_route_table.test-private-route-table.id}",
        "subnet_id": "\${aws_subnet.test-private-subnet-a.id}"
      },
      "test-route-table-association-private-subnet-c": {
        "route_table_id": "\${aws_route_table.test-private-route-table.id}",
        "subnet_id": "\${aws_subnet.test-private-subnet-c.id}"
      },
      "test-route-table-association-public-subnet-a": {
        "route_table_id": "\${aws_route_table.test-public-route-table.id}",
        "subnet_id": "\${aws_subnet.test-public-subnet-a.id}"
      },
      "test-route-table-association-public-subnet-c": {
        "route_table_id": "\${aws_route_table.test-public-route-table.id}",
        "subnet_id": "\${aws_subnet.test-public-subnet-c.id}"
      }
    },
    "aws_subnet": {
      "test-private-subnet-a": {
        "availability_zone": "ap-northeast-2a",
        "cidr_block": "10.0.101.0/24",
        "tags": {
          "CreatedBy": "cdktf",
          "Name": "test-private-subnet-a",
          "Project": "cdktf-for-vpc",
          "how": "it"
        },
        "vpc_id": "\${aws_vpc.test-vpc.id}"
      },
      "test-private-subnet-c": {
        "availability_zone": "ap-northeast-2c",
        "cidr_block": "10.0.103.0/24",
        "tags": {
          "CreatedBy": "cdktf",
          "Name": "test-private-subnet-c",
          "Project": "cdktf-for-vpc",
          "how": "it"
        },
        "vpc_id": "\${aws_vpc.test-vpc.id}"
      },
      "test-public-subnet-a": {
        "availability_zone": "ap-northeast-2a",
        "cidr_block": "10.0.1.0/24",
        "map_public_ip_on_launch": true,
        "tags": {
          "CreatedBy": "cdktf",
          "Name": "test-public-subnet-a",
          "Project": "cdktf-for-vpc",
          "how": "it"
        },
        "vpc_id": "\${aws_vpc.test-vpc.id}"
      },
      "test-public-subnet-c": {
        "availability_zone": "ap-northeast-2c",
        "cidr_block": "10.0.3.0/24",
        "map_public_ip_on_launch": true,
        "tags": {
          "CreatedBy": "cdktf",
          "Name": "test-public-subnet-c",
          "Project": "cdktf-for-vpc",
          "how": "it"
        },
        "vpc_id": "\${aws_vpc.test-vpc.id}"
      }
    },
    "aws_vpc": {
      "test-vpc": {
        "cidr_block": "10.0.0.0/16",
        "tags": {
          "CreatedBy": "cdktf",
          "Name": "test-vpc",
          "Project": "cdktf-for-vpc",
          "how": "it"
        }
      }
    }
  },
  "terraform": {
    "required_providers": {
      "aws": {
        "source": "aws",
        "version": "4.60.0"
      }
    }
  }
}"
`;
